build: add Meson support for pgroonga module

otegami opened the pull request:
GitHub: GH-490

We added pgroonga module to the meson build system like other modules did.
We will add simple tests. In the following PR, we will add full test step.

kou commented on meson_options.txt:
```diff
@@ -2,3 +2,15 @@ option('pg_config',
+       type: 'feature',
+       value: 'disabled',
+       description: 'Enable MessagePack support')
+option('msgpack_package_name',
+       type: 'string',
+       value: 'msgpack',
+       description: 'MessagePack package name (e.g. "msgpack-c" on macOS)')
```
We should detect it automatically like Groonga does.

otegami replied:
fix: [62dbf95](https://github.com/pgroonga/pgroonga/pull/750/commits/62dbf9598951c96f502fec7fc30176275fb70b0a) Sure.

kou commented on meson_options.txt:
```diff
@@ -2,3 +2,15 @@ option('pg_config',
        description: 'Path to pg_config')
+option('msgpack',
+       type: 'feature',
+       value: 'disabled',
```
```suggestion
       value: 'auto',
```

otegami replied:
fix: [8045319](https://github.com/pgroonga/pgroonga/pull/750/commits/8045319acc4501567ec2c399860fa47a3289f4d5)

kou commented on meson_options.txt:
```diff
@@ -2,3 +2,15 @@ option('pg_config',
+       description: 'MessagePack package name (e.g. "msgpack-c" on macOS)')
+option('xxhash',
+       type: 'feature',
+       value: 'disabled',
```
```suggestion
       value: 'auto',
```

kou commented on meson_options.txt:
```diff
@@ -2,3 +2,15 @@ option('pg_config',
        type: 'string',
        value: '',
        description: 'Path to pg_config')
+option('msgpack',
```
```suggestion
option('message_pack',
```

otegami replied:
fix: [8a354a1](https://github.com/pgroonga/pgroonga/pull/750/commits/8a354a119aebe923d0e883e33854ecdcfcd19fdb)

kou commented on meson.build:
```diff
@@ -185,3 +380,12 @@ pgroonga_database_update_sqls = files(
 install_data(pgroonga_database_update_sqls,
   install_dir: pg_sharedir / 'extension',
 )
+
+pgroonga_scripts = files(
+  'tools/pgroonga-primary-maintainer.sh',
+  'tools/systemd/pgroonga-generate-primary-maintainer-service.sh',
+  'tools/systemd/pgroonga-generate-primary-maintainer-timer.sh',
+)
+install_data(pgroonga_scripts,
+  install_dir: pg_bindir,
+)
```
Should we add this in this PR?

otegami replied:
fix: [be1e768](https://github.com/pgroonga/pgroonga/pull/750/commits/be1e768407d147cdc83e27fda911f89eb0919b93) We don't need them now. So I removed them.

kou commented on meson.build:
```diff
@@ -40,15 +43,83 @@ postgresql = declare_dependency(
 pgrn_module_args = {
   'c_args': pgrn_c_args,
-  'dependencies': [postgresql, groonga, libm],
+  'cpp_args': pgrn_cpp_args,
```
Is this correct? Is this for C++?

otegami replied:
fix: [4af310a](https://github.com/pgroonga/pgroonga/pull/750/commits/4af310ada03a8d28ced448a170ba3c0ea85f4f4d) Ahh I totally misunderstood it. Thank you so much! Now, we don't need it because we declare it as a dependecy.

kou commented on meson.build:
```diff
@@ -40,15 +43,83 @@ postgresql = declare_dependency(
+pgrn_dependencies = [postgresql, groonga, libm]
+
+pgrn_cpp_args = []
+if get_option('msgpack').enabled()
+  msgpack = dependency(get_option('msgpack_package_name'), required: true)
```
```suggestion
msgpack = dependency(get_option('msgpack_package_name'), required: get_option('msgpack'))
if msgpack.found()
```

kou commented on meson.build:
```diff
@@ -40,15 +43,83 @@ postgresql = declare_dependency(
+if get_option('msgpack').enabled()
+  msgpack = dependency(get_option('msgpack_package_name'), required: true)
+  pgrn_dependencies += msgpack
+  pgrn_cpp_args += '-DPGRN_HAVE_MSGPACK'
```
We should add a dependency for this: `pgrn_msgpack = declare_dependency(compile_args: ['-DPGRN_HAVE_MSGPACK'], dependencies: [msgpack])`

kou commented on meson.build:
```diff
@@ -40,15 +43,83 @@ postgresql = declare_dependency(
+endif
+
+if get_option('xxhash').enabled()
+  xxhash = declare_dependency(
+    dependencies: [dependency('libxxhash', required: true)],
+    sources: files('vendor/xxHash/xxhash.c'),
+    include_directories: include_directories('vendor/xxHash'),
+  )
```
We must use system xxHash or vendored xxHash instead of both of them.

kou reviewed (commented):
Can we run at least one test to check whether this works or not?

otegami commented:
Thank you for your review. I’ve addressed all of the review comments here.

> Can we run at least one test to check whether this works or not?

fix: I added a simple test. For testing, I temporarily installed PGroonga under system's PostgreSQL.

kou commented on meson.build:
```diff
@@ -40,15 +43,92 @@ postgresql = declare_dependency(
+    dependencies: [msgpack]
+  )
+  pgrn_dependencies += pgrn_msgpack
+  message('MessagePack: system')
```
We don't need this because Meson shows it: https://github.com/pgroonga/pgroonga/actions/runs/15867044950/job/44735803391?pr=750#step:7:27

```text
Run-time dependency msgpack found: YES 4.0.0
```

```suggestion
```

kou commented on meson.build:
```diff
@@ -40,15 +43,92 @@ postgresql = declare_dependency(
+  message('MessagePack: system')
+endif
+
+xxhash_option = get_option('xxhash')
+if not xxhash_option.disabled()
+  xxhash = dependency('libxxhash', required: false)
+  if xxhash.found()
+    pgrn_dependencies += xxhash
+    message('xxHash: system')
+  else
+    pgrn_xxhash = declare_dependency(
+      sources: files('vendor/xxHash/xxhash.c'),
+      include_directories: include_directories('vendor/xxHash'),
+    )
+    pgrn_dependencies += pgrn_xxhash
+    message('xxHash: vendored')
+  endif
+endif
```
```suggestion
xxhash = dependency('libxxhash', required: get_option('xxhash'))
if xxhash.found()
  pgrn_dependencies += xxhash
endif
```

otegami commented on .github/workflows/linux.yml:
```diff
@@ -107,16 +107,25 @@ jobs:
+          meson setup ../pgroonga.build \
+            -Dpg_config=${PG_CONFIG} \
+            -Dmessage_pack=enabled \
+            -Dxxhash=enabled
```
fix: [1c85305](https://github.com/pgroonga/pgroonga/pull/750/commits/1c853056fe8bbf41adecfa825e99028da76e5647)

otegami commented on meson.build:
```diff
@@ -40,15 +43,102 @@ postgresql = declare_dependency(
+  msgpack = dependency('msgpack', required: false)
+  if not msgpack.found()
+    msgpack = dependency('msgpack-c', required: false)
+  endif
```
fix: [cf39268](https://github.com/pgroonga/pgroonga/pull/750/commits/cf392681bf408376061d0c3538de59a2fc317c4b)

otegami commented on meson.build:
```diff
@@ -185,3 +399,23 @@ pgroonga_database_update_sqls = files(
+    '--bindir=' + pg_bindir,
+    '--dlpath=' + meson.current_build_dir(),
+    '--load-extension=pgroonga',
+    'column-name/japanese',
```
fix: [aba6db2](https://github.com/pgroonga/pgroonga/pull/750/commits/aba6db2c482e8af4dbb9461ee46cc9b62aea2c6b)

otegami commented:
Thank you for your review. I’ve addressed all of the review comments here.
